### ISSUE TITLE
Node's Indexer does not resume indexing after stoping the node

### ISSUE THREAD

**Issue by mariacarmina** [oceanprotocol/ocean-node#740]
**Describe the bug**
Node's Indexer does not resume indexing after stoping the node (with Ctrl + C or Cmd + C). The expected behaviour is to resume the last indexed block, not starting from deployer block. The issue is fixed if the startBlock is specified within RPCS env var, but it is not user intuitive/friendly to keep updating startBlock.

**To Reproduce**
1. Run Ocean Node on local with npm run start
2. Wait the node to sync
3. Stop the node
4. Restart node

**Expected behavior**
The expected behaviour is to resume the last indexed block, not starting from deployer block.

**Logs**
[ocean-node-op-sepolia-indexer.txt](https://github.com/user-attachments/files/17556839/ocean-node-op-sepolia-indexer.txt)

**Comment by md00ux** [comment]
- **ocean-node** container was started using docker compose 
- database type => typesense
```yaml
DB_TYPE: 'typesense'
DB_URL: "http://typesense:8108/?apiKey=xyz"
....
typesense:
    image: typesense/typesense:26.0
    container_name: typesense
    ports:
      - "8108:8108"
    networks:
      - backend
    volumes:
      - typesense-data:/data
    command: '--data-dir /data --api-key=xyz'
```

**Comment by paulo-ocean** [comment]
It looks to me that we could not get the `lastIndexedBlock` from DB for some reason.. that's the only explanation i can see by looking at the code.. if for some reason we fail to get that info, it will start from either `rpcDetails.startBlock` or `contractDeploymentBlock` (whatever is more up to date), but that is expected, as we cannot guess the last indexed block..
I would say the problem here was at DB/connection level

**Comment by mariacarmina** [comment]
> It looks to me that we could not get the `lastIndexedBlock` from DB for some reason.. that's the only explanation i can see by looking at the code.. and the logs... The expected behaviour is Ok.
> 
> But if for some reason we fail to get that info, it will start from either `rpcDetails.startBlock` or `contractDeploymentBlock` (whatever is more up to date), but that is expected, as we cannot guess the last indexed block.. The `crawlingStartBlock` variable is calculated from these
> 
> I would say the problem here was at DB/connection level. Maybe we had no data there on the restart? We lost it?
> 
> @md00ux can we have the full logs?
> 
> 
> 
> We can easily improve the logging, so we know exactly what we are reading:
> 
> ```
> 
> const lastIndexedBlock = await getLastIndexedBlock()
> 
>       const networkHeight = await getNetworkHeight(provider)
> 
>       const startBlock =
> 
>         lastIndexedBlock && lastIndexedBlock > crawlingStartBlock
> 
>           ? lastIndexedBlock
> 
>           : crawlingStartBlock
> 
> 
> 
>       INDEXER_LOGGER.logMessage(
> 
>         `network: ${rpcDetails.network} Start block ${startBlock} network height ${networkHeight}`,
> 
>         true
> 
>       )
> 
> ```
> 
> We should add the `lastIndexedBlock` retrieved to the log message. That will clarify it

Hello @paulo-ocean!

Thank you for jumping on this issue.
Re-trying the same scenario today, and it couldn't be reproduced and maybe it was a database connection issue, as we discussed, when retrieving the last indexed block.

I have a suggestion regarding this issue, does it help to have a retry mechanism if db connection breaks or a known db error occurs to assert it. We can implement a while with number of retries set on 2 for example, because we do not want to loop too much for retry if db connection is down for good for example.
Also I am not sure if retry mechanisms are indicated for working with threads, but that section when retrieving the last indexed block is a safe-thread zone, so it shouldn't represent a problem.

Let me know what you think.

**Comment by paulo-ocean** [comment]
Hi @mariacarmina 
Have a look at this PR, to imporve logging and a couple improvements on other functions related with updating/getting the block.
About the retry mechanism, we already have one before starting to index anything.. it only starts with a proper RPC and DB connection.. Apart from this, we also have a ReindexChain command (have another PR to improve that with an optional block number as well).
Now, If the DB connection is lost somewhere in the "middle" of the operations, the "retry" will automatically happen at the next run of the while() loop.. So, we already have that

**Comment by md00ux** [comment]
I tested the same scenario and was not able to reproduce the behavior.
- started the ocean-node and waited until the blockchain was fully indexed.
- stopped the ocean-node container and started it again.

The logs show that indexing resumed from the last indexed block:

```shell
2024-10-30T07:48:19.794Z info: OCEANNODE:	🌊 [ Starting Ocean Node ]
2024-10-30T07:48:19.795Z info: CONFIG:	✅ INDEXER_NETWORKS is not defined, running Indexer with all supported networks defined in RPCS env variable ...
2024-10-30T07:48:19.809Z info: CONFIG:	✅ Starting node with peerID: 16Uiu2HAmMwDUBReP5EXXgB3cTU2SJkxW2LKMShog8N1UC7GRMXfk
2024-10-30T07:48:19.812Z warn: CONFIG:	⚠ ️ Invalid or missing "AUTHORIZED_DECRYPTERS" env variable...
2024-10-30T07:48:19.816Z warn: CONFIG:	⚠ ️ Invalid or missing "ALLOWED_VALIDATORS" env variable...
2024-10-30T07:48:19.818Z warn: CONFIG:	⚠ ️ Invalid or missing "P2P_BOOTSTRAP_NODES" env variable...
2024-10-30T07:48:19.820Z warn: CONFIG:	⚠ ️ Invalid or missing "P2P_ANNOUNCE_ADDRESSES" env variable...
2024-10-30T07:48:19.820Z warn: CONFIG:	Missing "FEE_AMOUNT" env variable. Will use defaults...
2024-10-30T07:48:19.823Z warn: CONFIG:	Missing "FEE_TOKENS" env variable. Will use defaults...
2024-10-30T07:48:19.825Z warn: CONFIG:	⚠ ️ Invalid or missing "OPERATOR_SERVICE_URL" env variable...
2024-10-30T07:48:19.827Z warn: CONFIG:	⚠ ️ Invalid or missing "ALLOWED_ADMINS" env variable...
2024-10-30T07:48:19.828Z warn: CONFIG:	Missing "MAX_REQ_PER_SECOND" env variable. Will use defaults...
2024-10-30T07:48:19.828Z warn: CONFIG:	⚠ ️ Invalid or missing "RATE_DENY_LIST" env variable...
2024-10-30T07:48:19.828Z warn: CONFIG:	⚠ ️ Invalid or missing "UNSAFE_URLS" env variable...
2024-10-30T07:48:21.283Z info: OCEANNODE:	Codebase hash: 2847dd48e570be7dabb33ff65171de11bdf22ed201b37bcb1ea549cf4cfbcd54
2024-10-30T07:48:21.283Z info: DATABASE:	Nonce Database initiated with SQLite provider
2024-10-30T07:48:21.298Z warn: DATABASE:	Property "LOG_DB" is set to "false". This means logs will NOT be saved to database!
2024-10-30T07:48:21.486Z info: P2P:	Binding P2P sockets to IPV4
2024-10-30T07:48:21.487Z info: P2P:	Binding P2P sockets to IPV6
2024-10-30T07:48:21.487Z info: P2P:	Enabling UPnp discovery
2024-10-30T07:48:21.488Z info: P2P:	Enabling AutoNat service
2024-10-30T07:48:21.488Z info: P2P:	Enabling P2P Transports: websockets, tcp, circuitRelay
2024-10-30T07:48:21.636Z info: OCEANNODE:	OCEANNODE:	HTTP port: 8000
2024-10-30T07:48:21.978Z info: INDEXER:	Blockchain connection succeffully established!
2024-10-30T07:48:21.983Z info: INDEXER:	Starting worker for network op-sepolia with {"rpcDetails":{"rpc":"<redacted>","chainId":11155420,"network":"op-sepolia","chunkSize":2000}}
(node:17) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
    at emitExperimentalWarning (node:internal/util:264:11)
    at ModuleLoader.jsonStrategy (node:internal/modules/esm/translators:456:3)
    at callTranslator (node:internal/modules/esm/loader:272:14)
    at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:278:30)
2024-10-30T07:48:24.028Z info: CONFIG:	✅ INDEXER_NETWORKS is not defined, running Indexer with all supported networks defined in RPCS env variable ...
secp256k1 unavailable, reverting to browser version
2024-10-30T07:48:25.274Z info: CONFIG:	✅ INDEXER_NETWORKS is not defined, running Indexer with all supported networks defined in RPCS env variable ...
2024-10-30T07:48:25.286Z info: DATABASE:	Nonce Database initiated with SQLite provider
2024-10-30T07:48:25.295Z warn: DATABASE:	Property "LOG_DB" is set to "false". This means logs will NOT be saved to database!
2024-10-30T07:48:25.638Z info: INDEXER:	INDEXER:	network: op-sepolia Start block 19235936 network height 19235982
2024-10-30T07:48:25.638Z info: network: op-sepolia processing 46 blocks ...
2024-10-30T07:48:25.792Z info: INDEXER:	INDEXER:	New last indexed block : 19235982
2024-10-30T07:48:56.092Z info: INDEXER:	INDEXER:	network: op-sepolia Start block 19235982 network height 19235997
2024-10-30T07:48:56.093Z info: network: op-sepolia processing 15 blocks ...
2024-10-30T07:48:56.265Z info: INDEXER:	INDEXER:	New last indexed block : 19235997
2024-10-30T07:49:00.637Z info: P2P:	Trying to punch a hole using UPNP
2024-10-30T07:49:26.551Z info: INDEXER:	INDEXER:	network: op-sepolia Start block 19235997 network height 19236012
2024-10-30T07:49:26.551Z info: network: op-sepolia processing 15 blocks ...
2024-10-30T07:49:26.706Z info: INDEXER:	INDEXER:	New last indexed block : 19236012
```
